pkg/internal/managedappcredential/manager_ensure.go

```diff
@@ -0,0 +1,199 @@
+		if err := m.runGarbageCollection(ctx, oldParentUser, appCredential.id, false); err != nil {
+			m.logger.Error(err, "could not clean up application credential(s) as the owning user has changed and information about owning user might be stale")
+		}
+	}
+
+	// In case the application credential usage is disabled, it is tried to
+	// clean up old application credentials before aborting.
+	if !m.config.Enabled {
+		if !appCredentialExists {
+			return m.removeApplicationCredentialStore(ctx, appCredential.secret)
+		}
+
+		if oldParentUserUsable {
+			if err := m.runGarbageCollection(ctx, oldParentUser, "", true); err != nil {
+				return err
+			}
+		}
+
+		return m.removeApplicationCredentialStore(ctx, appCredential.secret)
+	}
+
+	// If the new parent user is itself an application credential, then no managed
+	// application credential can be used. It will be tried to clean up with the
+	// old user if this one is usuable.
+	if newParentUser.isApplicationCredential() {
+		if appCredentialExists {
+			if err := m.runGarbageCollection(ctx, oldParentUser, appCredential.id, true); err != nil {
+				return err
+			}
+		}
+
+		return m.removeApplicationCredentialStore(ctx, appCredential.secret)
+	}
+
+	// No application credential exists so a new one need to be created.
+	if !appCredentialExists {
+		if err := m.runGarbageCollection(ctx, newParentUser, "", false); err != nil {
+			return err
+		}
+
+		newAppCredential, err := m.createApplicationCredential(ctx, newParentUser)
+		if err != nil {
+			return err
+		}
+
+		return m.storeApplicationCredential(ctx, newAppCredential, newParentUser)
+	}
+
+	if err := m.runGarbageCollection(ctx, newParentUser, appCredential.id, false); err != nil {
+		return err
+	}
+
+	// In case the parent users have changed a new app credential need to be created
+	// and it is tried to remove the one managed by the old user.
+	if parentChanged {
+		newAppCredential, err := m.createApplicationCredential(ctx, newParentUser)
+		if err != nil {
+			return err
+		}
+
+		// Try to delete the old application credential owned by the old user.
+		// This might not work as the information about this user could be stale,
+		// because the user credentials are rotated, the user is not associated to
+		// Openstack project anymore or it is deleted.
+		if err := oldParentUser.identityClient.DeleteApplicationCredential(ctx, oldParentUser.id, appCredential.id); err != nil {
+			m.logger.Error(err, "could not delete application credential as the owning user has changed and information about owning user might be stale")
+		}
+
+		return m.storeApplicationCredential(ctx, newAppCredential, newParentUser)
+	}
+
+	if m.isExpired(appCredential) {
+		newAppCredential, err := m.createApplicationCredential(ctx, newParentUser)
+		if err != nil {
+			return err
+		}
+
+		return m.storeApplicationCredential(ctx, newAppCredential, newParentUser)
+	}
+
+	return m.storeApplicationCredential(ctx, appCredential, newParentUser)
+}
```

Comment:
I get the feeling that the code can be condensed a bit more e.g.
```
if parentChanged || m.isExpired(appCredential) || !appCredentialExists {
		// Try to delete the old application credential owned by the old user.
		// This might not work as the information about this user could be stale,
		// because the user credentials are rotated, the user is not associated to
		// Openstack project anymore or it is deleted.
		if parentChanged{
				if err := oldParentUser.identityClient.DeleteApplicationCredential(ctx, oldParentUser.id, appCredential.id); err != nil {
			m.logger.Error(err, "could not delete application credential as the owning user has changed and information about owning user might be stale")
		         }
                }

		newAppCredential, err := m.createApplicationCredential(ctx, newParentUser)
		if err != nil {
			return err
		}
		if err :=  m.storeApplicationCredential(ctx, newAppCredential, newParentUser){
                    return err
                 }
}
if err := m.runGarbageCollection(ctx, newParentUser, appCredential.id, false); err != nil {
		return err
}

```
It is easy to miss a case anyway so I am not sure how big of an improvement it is.